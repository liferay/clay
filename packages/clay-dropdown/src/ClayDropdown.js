import 'clay-icon';
import './ClayDropdownBase';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import {itemsValidator, preferredAlign} from './validators';
import templates from './ClayDropdown.soy.js';

/**
 * Implementation for Metal Clay Dropdown.
 * @extends ClayComponent
 */
class ClayDropdown extends ClayComponent {
	/**
	 * @inheritDoc
	 */
	attached() {
		this.addListener('toggle', this._defaultToggle, true);
	}

	/**
	 * Toggles the dropdown, closing it when open or opening it when closed.
	 * @protected
	 */
	_defaultToggle() {
		this.expanded = !this.expanded;
	}

	/**
	 * Handles footer button click.
	 * @param {!Event} event
	 * @protected
	 */
	_handleButtonClick(event) {
		this.emit('buttonClicked', event);
	}

	/**
	 * Continues the propagation of the item clicked event
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleItemClick(event) {
		return !this.emit({
			data: event.data,
			name: 'itemClicked',
			originalEvent: event,
		});
	}

	/**
	 * Continues the propagation of the toggle event.
	 * @param {?Event} event
	 * @return {Boolean} If the event has been prevented or not.
	 */
	toggle(event) {
		if (event) {
			event.preventDefault();
		}

		return !this.emit({
			name: 'toggle',
		});
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayDropdown.STATE = {
	/**
	 * Button configuration to place a button at dropdown footer.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	button: Config.shapeOf({
		label: Config.string().required(),
		style: Config.oneOf(['primary', 'secondary']),
		type: Config.oneOf(['button', 'reset', 'submit']),
	}),

	/**
	 * Flag to indicate if menu is disabled
	 * @default false
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?bool}
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to indicate if menu is expanded.
	 * @default false
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?bool}
	 */
	expanded: Config.bool().value(false),

	/**
	 * Id to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	id: Config.string(),

	/**
	 * Icon to use in the trigger.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	icon: Config.string(),

	/**
	 * List of menu items.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {!Array}
	 */
	items: itemsValidator.required(),

	/**
	 * Position in which item icon will be placed. Needed if any item has icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	itemsIconAlignment: Config.oneOf(['left', 'right']),

	/**
	 * Label of the trigger button.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(html|string|undefined)}
	 */
	label: Config.any(),

	/**
	 * The preferred alignment for the dropdown content.
	 * @default BottomLeft
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	preferredAlign: preferredAlign.value('BottomLeft'),

	/**
	 * Flag to indicate if menu has a search field and search through elements
	 * is possible.
	 * @default false
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?bool}
	 */
	searchable: Config.bool().value(false),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {!string}
	 */
	spritemap: Config.string().required(),

	/**
	 * Style of the trigger button.
	 * @default unstyled
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string}
	 */
	style: Config.oneOf(['link', 'primary', 'secondary', 'unstyled']).value(
		'unstyled'
	),

	/**
	 * CSS classes to be applied to the trigger element.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	triggerClasses: Config.string(),

	/**
	 * The title attribute of the trigger element.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	triggerTitle: Config.string(),

	/**
	 * Size of the trigger button.
	 * @default undefined
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?(string|undefined)}
	 */
	triggerSize: Config.oneOf(['sm']),
};

defineWebComponent('clay-dropdown', ClayDropdown);

Soy.register(ClayDropdown, templates);

export {ClayDropdown};
export default ClayDropdown;
