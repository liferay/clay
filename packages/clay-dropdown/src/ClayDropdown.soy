{namespace ClayDropdown}

/**
 * This renders a dropdown.
 */
{template .render}
	{@param items: list<?>}
	{@param spritemap: string}
	{@param? button: ?}
	{@param? elementClasses: string}
	{@param? expanded: bool}
	{@param? handleButtonClick_: any}
	{@param? handleItemClick_: any}
	{@param? icon: string}
	{@param? id: string}
	{@param? itemsIconAlignment: string}
	{@param? label: html|string}
	{@param? searchable: bool}
	{@param? style: string}
	{@param? triggerClasses: string}

  {let $trigger kind="html"}
    {if $icon}
      {call ClayIcon.render}
        {param spritemap: $spritemap /}
        {param symbol: $icon /}
      {/call}
    {/if}

    {if $label}
      {$label}
    {/if}

    {if not $icon}
      {call ClayIcon.render}
        {param spritemap: $spritemap /}
        {param symbol: 'caret-bottom' /}
      {/call}
    {/if}
  {/let}

	{call ClayDropdownBase.render}
    {param button: $button /}
    {param elementClasses: $elementClasses /}
    {param events: [
      'buttonClicked': $handleButtonClick_,
      'itemClicked': $handleItemClick_
    ]/}
    {param expanded: $expanded /}
    {param id: $id /}
    {param items: $items /}
    {param itemsIconAlignment: $itemsIconAlignment /}
    {param label: $trigger /}
    {param ref: 'dropdown' /}
    {param searchable: $searchable /}
    {param spritemap: $spritemap /}
    {param style: $style /}
    {param triggerClasses: $triggerClasses /}
  {/call}
{/template}
