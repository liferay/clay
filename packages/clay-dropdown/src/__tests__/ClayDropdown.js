import ClayDropdown from '../ClayDropdown';

let clayDropdown;

let items = [
	{
		href: 'item1url',
		label: 'Item 1',
	},
	{
		href: 'item2url',
		label: 'Item 2',
	},
	{
		active: true,
		href: 'item1ur3',
		label: 'Item 3',
	},
	{
		disabled: true,
		href: 'item1url',
		label: 'Item 4',
	},
	{
		inputName: 'checkbox5name',
		inputValue: 'checkbox5value',
		label: 'Item 5',
		type: 'checkbox',
	},
	{
		checked: true,
		inputName: 'checkbox6name',
		inputValue: 'checkbox7value',
		label: 'Item 7',
		type: 'checkbox',
	},
	{
		disabled: true,
		inputName: 'checkbox7name',
		inputValue: 'checkbox7value',
		label: 'Item 7',
		type: 'checkbox',
	},
	{
		items: [
			{
				label: 'Item 8',
				inputValue: 'radio8value',
			},
			{
				checked: true,
				label: 'Item 9',
				inputValue: 'radio9value',
			},
			{
				disabled: true,
				label: 'Item 10',
				inputValue: 'radio10value',
			},
		],
		inputName: 'radioname',
		label: 'Radio Group',
		separator: true,
		type: 'radiogroup',
	},
];

describe('ClayDropdown', function() {
	afterEach(() => {
		if (clayDropdown) {
			clayDropdown.dispose();
		}
	});

	it('should render a dropdown with items', () => {
		clayDropdown = new ClayDropdown({
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown with classes', () => {
		clayDropdown = new ClayDropdown({
			elementClasses: 'my-custom-class',
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown with id', () => {
		clayDropdown = new ClayDropdown({
			id: 'myId',
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown with trigger custom classes', () => {
		clayDropdown = new ClayDropdown({
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
			triggerClasses: 'my-custom-class',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render an expanded dropdown', () => {
		clayDropdown = new ClayDropdown({
			expanded: true,
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown with an icon trigger', () => {
		clayDropdown = new ClayDropdown({
			icon: 'list',
			items: items,
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a searchable dropdown', () => {
		clayDropdown = new ClayDropdown({
			items: items,
			label: 'Trigger',
			searchable: true,
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown and emit an event on button click', () => {
		clayDropdown = new ClayDropdown({
			button: {
				label: 'Button',
				style: 'primary',
			},
			items: [
				{
					href: 'item1url',
					label: 'Item 1',
				},
			],
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		const spy = jest.spyOn(clayDropdown.refs.dropdown, 'emit');

		const element =
			clayDropdown.refs.dropdown.refs.portal.refs.dropdownButton.element;
		element.click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith('buttonClicked', expect.any(Object));
	});
});
