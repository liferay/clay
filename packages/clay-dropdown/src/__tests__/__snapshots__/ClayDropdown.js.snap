// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayDropdown should render a dropdown with an icon trigger 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-list">
      <title>list</title>
      <use xlink:href="icons.svg#list"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;

exports[`ClayDropdown should render a dropdown with classes 1`] = `
<div class="dropdown my-custom-class">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled" type="button">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <title>caret-bottom</title>
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;

exports[`ClayDropdown should render a dropdown with id 1`] = `
<div class="dropdown" id="myId">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled" type="button">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <title>caret-bottom</title>
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;

exports[`ClayDropdown should render a dropdown with items 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled" type="button">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <title>caret-bottom</title>
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;

exports[`ClayDropdown should render a dropdown with trigger custom classes 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn my-custom-class btn-unstyled" type="button">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <title>caret-bottom</title>
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;

exports[`ClayDropdown should render a searchable dropdown 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled" type="button">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <title>caret-bottom</title>
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <form>
    <div class="dropdown-section">
      <div class="input-group input-group-sm">
        <div class="input-group-item">
          <input aria-label="Search for..." class="form-control input-group-inset input-group-inset-after" placeholder="Search for..." type="text" ref="searchInput">
          <span class="input-group-inset-item input-group-inset-item-after">
            <button class="btn btn-link" aria-label="search" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-search">
                <title>search</title>
                <use xlink:href="icons.svg#search"></use>
              </svg>
            </button>
          </span>
        </div>
      </div>
    </div>
  </form>
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;

exports[`ClayDropdown should render an expanded dropdown 1`] = `
<div class="dropdown show">
  <button aria-expanded="true" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled" type="button">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <title>caret-bottom</title>
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu show" ref="menu" style="top: 0px; left: 0px;">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class=" dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class=" dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 5</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item active">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
    <li class="dropdown-item">
      <div class="custom-control custom-checkbox custom-control-inline">
        <label>
          <input disabled="disabled" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
          <span class="custom-control-label">
            <span class="custom-control-label-text">Item 7</span>
          </span>
        </label>
      </div>
    </li>
  </ul>
  <div class="dropdown-subheader" role="presentation">Radio Group</div>
  <div aria-label="Radio Group" role="radiogroup">
    <ul class="list-unstyled">
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input class="custom-control-input" value="radio8value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 8</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item active">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input checked="" class="custom-control-input" value="radio9value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 9</span>
            </span>
          </label>
        </div>
      </li>
      <li class="dropdown-item">
        <div class="custom-control custom-radio custom-control-inline">
          <label>
            <input disabled="disabled" class="custom-control-input" value="radio10value" ref="input" type="radio" role="radio">
            <span class="custom-control-label">
              <span class="custom-control-label-text">Item 10</span>
            </span>
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div aria-hidden="true" class="dropdown-divider" role="presentation"></div>
</div>
`;
